openedx/core/djangoapps/bookmarks/tasks.py

```diff
@@ -0,0 +1,155 @@
+"""
+Tasks for bookmarks.
+"""
+import logging
+
+from django.db import transaction
+
+from celery.task import task  # pylint: disable=import-error,no-name-in-module
+from opaque_keys.edx.keys import CourseKey
+from xmodule.modulestore.django import modulestore
+
+log = logging.getLogger('edx.celery.task')
+
+
+def _calculate_course_xblocks_data(course_key):
+    """
+    Fetch data for all the blocks in the course.
+
+    This data consists of the display_name and path of the block.
+    """
+    with modulestore().bulk_operations(course_key):
+
+        course = modulestore().get_course(course_key, depth=None)
+        blocks_info_dict = {}
+
+        # Collect display_name and children usage keys.
+        blocks_stack = [course]
+        while blocks_stack:
+            current_block = blocks_stack.pop()
+            children = current_block.get_children() if current_block.has_children else []
+            usage_id = unicode(current_block.scope_ids.usage_id)
+            block_info = {
+                'usage_key': current_block.scope_ids.usage_id,
+                'display_name': current_block.display_name,
+                'children_ids': [unicode(child.scope_ids.usage_id) for child in children]
+            }
+            blocks_info_dict[usage_id] = block_info
+
+            # Add this blocks children to the stack so that we can traverse them as well.
+            blocks_stack.extend(children)
+
+    # Set children
+    for block in blocks_info_dict.values():
+        block.setdefault('children', [])
+        for child_id in block['children_ids']:
+            block['children'].append(blocks_info_dict[child_id])
+        block.pop('children_ids', None)
+
+    # Calculate paths
+    def add_path_info(block_info, current_path):
+        """Do a DFS and add paths info to each block_info."""
+
+        block_info.setdefault('paths', [])
+        block_info['paths'].append(current_path)
+
+        for child_block_info in block_info['children']:
+            add_path_info(child_block_info, current_path + [block_info])
+
+    add_path_info(blocks_info_dict[unicode(course.scope_ids.usage_id)], [])
+
+    return blocks_info_dict
+
+
+def _paths_from_data(paths_data):
+    """
+    Construct a list of paths from path data.
+    """
+    from .models import PathItem
```

Comment:
This import shows some circular dependencies, can we move `PathItem` to `init` to avoid such imports.